### DISCUSSION TITLE
webrtc video encode of webgpu output

### DISCUSSION THREAD

**Discussion by unicomp21** [gpuweb/gpuweb#2673] [Ideas/Feedback]
why not allow video encoding of the webgpu framebuffer to webrtc video stream?  don't most gpu's now support video encode?  wouldn't this be huge for collaboration when using CAD tools, etc. peer to peer?  and also be useful for server side rendering when using deno?

**Comment by Kangz** [comment]
AFAIK WebRTC can encode what's rendered to a canvas. So you should be able to render to a canvas using WebGPU and then put that canvas in a WebRTC stream. I'm not sure what's the implementation status in the various browsers, but eventually this should work.

**Reply by unicomp21** [reply]
yeah, I'm aware, but how efficient will the user space draw call be per frame?

**Reply by Kangz** [reply]
User space draw call? Draws you do in render passes to draw to the canvas should be as efficient as draws you make to a canvas that's not hooked to WebRTC.

**Reply by unicomp21** [reply]
My question relates to drawing on a canvas, which is then video encoded vs drawing to a framebuffer which is then video encoded.  Is one more efficient that the other?

**Reply by Kangz** [reply]
There is no way to have a "framebuffer" that gets video encoded on the Web platform. The only way to do this is by drawing to a canvas. Browsers should try to optimize that path and use hardware video encoding when they can but that depends on the OS / video codec / implementation status.

But with video conferencing sites like Zoom / Teams / Meet using WebGL (and in the future WebGPU?) for effects, browsers are incentivized to make that path efficient.
